torch_xla/experimental/spmd_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,107 @@
+from typing import (Any, Callable, Optional, Union)
+
+import torch
+import torch.nn as nn
+from torch._prims_common import TensorLike, TensorSequenceType
+
+import numpy as np
+
+import torch_xla
+import torch_xla.distributed.spmd as spmd
+
+def _prepare_spmd_partition_spec(param):
+  partition_spec = [None] * len(param.shape)
+  # Skip scalar tensors.
+  if len(partition_spec) == 0:
+    return partition_spec
+
+  # Only shard the 0th dimension of the parameter according to the
+  # fsdp axis of the mesh.
+  # TODO: should we shard on the maximal dim for param? Then we need
+  # another helper for the output.
+  partition_spec[0] = "fsdp"
+  return tuple(partition_spec)
+
+
+class SpmdFullyShardedDataParallel(nn.Module):
+
+  def __init__(self, module: nn.Module, mesh: spmd.Mesh, shard_output:Optional[Callable] = None):
```

Comment:
do we need to take mesh from caller? If it is FSDP, would it be possible to just assume we just shard at 0th dimension for all devices if mesh is not provided?

Reply:
The thing is that the mesh will need to share with the dataloader. And a global mesh is a must for SPMD. I was thinking maybe we can have an API to set a global mesh. @jonb377 @yeounoh